### ISSUE TITLE
Option to delay engineer responses until beginning of next business day

### ISSUE THREAD

**Issue by mjtko** [alces-software/alces-flight-center#740]
Sometimes we do work on cases outside of what are considered business hours by Flight Center (😱).  When work like this is performed, we'd like to be able to write a comment but not send the email notification to the customer.  Ideally, the Slack notification should still be sent at the time the work is done, even more ideally containing a note that the email notification is being deferred.

Possible implementation: check-box ("Defer response until next business day") shown outside of business hours (maybe always?), that is checked by default outside of business hours (would be unchecked during business hours if we always show it).

Responses should not all be sent at exactly 9am.  Thought: 9am + random(40) minutes.

To prevent double-replying, it should be obvious within the Flight Center UI that the comment/change is pending.  i.e. Obviously highlight comments/changes that have not yet been dispatched to the customer.

Consider: this should apply to any changes performed to a case that trigger an email notification to the customer, not just comments, though comments/resolved status notifications are the most important thing to handle.

**Comment by jamesremuscat** [comment]
Implementation note: Rails' `deliver_later` method includes a `wait_until` option that might be of use here! https://api.rubyonrails.org/classes/ActionMailer/MessageDelivery.html#method-i-deliver_later

**Comment by jamesremuscat** [comment]
To clarify what is being asked for here, when a case is modified after-hours, are we:
1. Applying the changes immediately, but postponing the email notification until business hours;
2. Recording the changes as pending, and applying them during business hours?

I ask because your comment about preventing double-replying confused me; if we're doing the former then the original reply will be there on the case, giving no higher risk of double-replying than at present. The latter will be significantly more work.

**Comment by jamesremuscat** [comment]
@mjtko says "the latter, i.e. the significantly more work one :slightly_smiling_face:"

Further thoughts and questions to come.

**Comment by jamesremuscat** [comment]
Discussion and questions at https://docs.google.com/document/d/1VLAa1ONGwqFMuOleBMKkUwMLX-6NkoIzPiP-Z82LZe4.

**Comment by mjtko** [comment]
With regard to the question regarding interleaved customer updates to a case when pending updates have been recorded, of the existing suggestions my favourite is:

> Should we prevent further updates to cases with pending changes? If so, how should this be communicated to the customer?

This would be a pragmatic first attempt at solving this issue and I'd accept the comment box being removed and a message shown to the effect of "An engineer has performed work on this case and has drafted a case update that will be published shortly."

Having said that, it's quite restrictive if a customer wants to add further details etc. Could we do:

1. Engineer creates update which is scheduled as outlined.
2. If a customer adds a comment, suspend engineer update and notify support team (Slack & email).
3. Engineer manually chooses either to proceed with update or to cancel the update and create a new one.

Probably worth showing a similar message to above, but with the comment box still shown -- I would retain the message until the engineer has decided whether to proceed or cancel the update.

> (Suggest: an alert at the top of the page listing the changes to be made, and probably the time they will be applied as well.)

Could this be displayed as a card on the case rather than as a flash/alert?

I feel that we might have to iterate on this a bit in order to get the desired UX but, in general, the suggestions in the document are sound and would be a good first shot at this functionality.

I have some further comments on the suggestions pertaining to #843 and will comment there re those.

**Comment by mjtko** [comment]
I couldn't see explicit mention of this in the document, so to reiterate:

If an update is made outside business hours it should be *optional* for the update to be deferred (any longer than the [now/then] usual buffer time) with the default set as "defer until next business day".

If we go down the route of always showing the possibility of deferment (which I don't think is necessarily a bad idea, there could well be use cases where we don't want to send the update until the following biz day), then the default should be not to defer until the next business day when displaying the case form during business hours.

Alternative implementation would be to always defer according to business hour rules, but have a "send update now" button to allow the engineer to override the schedule.  This capability might be useful in any case, should the engineer change their minds or want to send a "buffered" update more rapidly.

**Comment by jamesremuscat** [comment]
I like the latter option - by default this will give the safety of the five-minute buffer, but allow overriding when it's needed.

I can think of a couple of ways of representing this in the UI:

- An "apply immediately" checkbox somewhere on the page - but it's likely to end up quite far away from most of the submit buttons so likely to get forgotten about
- Giving an "apply now" option on buffered changes (e.g. on the case "history" card) - would require an extra click/page load cycle, but would also satisfy the "engineer manually chooses to proceed with update" requirement as well, so I think I prefer it.

I wonder if we then need to limit the number of pending changes to 1 or not? It might get confusing if multiple changes are stacked up, but it's simple enough to apply the "new changes mark all pending ones as held and they don't get automatically applied" rule.

wrt displaying pending changes - happy to move the details to a card, but I think an alert at the top of the page (as we do for change requests) would also be useful as otherwise I can see people not scrolling down far enough to see the case history before changing things, not realising that there were pending updates.

**Comment by mjtko** [comment]
Yup, I agree with all of the above.  The concept of limiting pending changes to 1 seems sensible too.
